src/profilers.cpp

```diff
@@ -2,19 +2,369 @@
+    @param regex           Regular expression provided by the user 
+                           to parse the output from the metric command
+    @param server_buffer   Size of the connection buffer to the server        
+  */
+
+  MetricReader::MetricReader(std::string metric_command, 
+                            std::string metric_name, 
+                            int freq,
+                            std::string regex,
+                            unsigned int server_buffer){
+    
+    this->metric_command = metric_command;
+    this->metric_name = metric_name;
+    this->freq = freq;
+    this->regex = regex;
+    this->server_buffer = server_buffer;
+    this->name = "MetricReader";
+
+  }
+
+  std::string MetricReader::get_name() {
+    return this->name;
+  }
+
+  void MetricReader::start(pid_t pid,
+                   ServerConnInstrs &connection_instrs,
+                   fs::path result_out,
+                   fs::path result_processed,
+                   bool capture_immediately) {
+
+    this->result_out = result_out;
+    std::string instrs = connection_instrs.get_instructions(this->get_thread_count());
+
+    fs::path stderr_metric_command;
+    stderr_metric_command = result_out / "metric_command_stderr.log";
+
+    const int ERROR_STDERR = 201;
+    const int ERROR_STDOUT_DUP2 = 202;
+    const int ERROR_STDERR_DUP2 = 203;
+    const int ERROR_NO_NUMBER_REGEX = 204;
+    const int ERROR_TOO_MANY_NUMBERS_REGEX = 205;
+    const int ERROR_CONVERSION_TO_FLOAT = 206;
+    const int ERROR_PARSING_CONNECTION_INSTRS = 207;
+    
+    
+    std::vector<std::string> parts = boost::program_options::split_unix(this->metric_command);
+
+    const char *path = parts[0].c_str();
+    char *parts_c_str[parts.size() + 1];
+
+    for (int i = 0; i < parts.size(); i++) {
+      parts_c_str[i] = (char *)parts[i].c_str();
+    }
+    parts_c_str[parts.size()] = NULL;
+
+
+    pid_t forked_metric_profiler = fork(); //separate process for profiler, parent handles errors
+
+    if(forked_metric_profiler == 0){
+      int code_metric_exec = 0;
+
+      int status;
+      pid_t command_status = waitpid(pid, &status, WNOHANG); //check if profiled command has finished executing
+      const int interval_ns = 1000 * 1000000 / this->freq;
+
+      while (command_status > 0){
+
+        struct timespec start, end;
+        clock_gettime(CLOCK_MONOTONIC, &start);
+
+        int pipe_fd[2];
+
+        if (pipe(pipe_fd) == -1) {
+          print("Could not establish communication pipe between metric-exec "
+                "and metric-read in \"" + this->get_name() + "\"! Terminating "
+                "the profiled command wrapper.", true, true);
+          kill(pid, SIGTERM);
+          return;
+        }
+
+        pid_t forked_metric_exec = fork();
+        if (forked_metric_exec == 0){
+          close(pipe_fd[0]);
+
+          fs::current_path(result_processed);
```

Comment:
There's no need of changing the working directory in your class (unlike ```Perf```).

Reply:
Done :+1: